Feature/buildtool server side nbt

xavierh opened the pull request:
Closes #487 Build tool: should use server side nbt files 
Closes #486 Buildtool: custom Huts 

# Changes proposed in this pull request:
- Add a UUID for the server
- Load the schematic when the world is loading (so we know where to look for schematics)
- Clear separation between huts and decorations
- Allow multiple subdirectories for schematics
- Override jar  huts and decorations provided by the jar file
- Allow to create new styles without modifying the jar file,

## Schematic loading
A schematic file is loading from the  - get the schematic in this order:
   - from the schematic directory
   - from the jar

On the client (WindowBuildTool):
 - Compare the md5 hash with the one from the server
 - Request the schematic from the server if the file is not found or the md5 does not match
 - Save it in the cache directory

The location of the schematic directory depends of the type of game:
- On the logical server: **World Directory**/minecolonies/schematics
- On the logical client: ${HOME}/.minecraft/minecolonies/**Server UUID**/schematics (on linux)

## Re-organise the schematics
The schematics are clearly separated between **huts** and **decorations** using 2 different subdirectory of the schematics folder.
- huts schematic in the decoration folder will be loaded as decoration only
- Other directories  in the schematics folder are still usable by the mod but will not be available to the BuildTool (Ex: miner's schematics and SupplyChest)

 - huts/
   - stone
     - Builder1.nbt
     - ...
     - WareHouse5.nbt
   - wooden
     - Builder1.nbt
     - ...
     - WareHouse5.nbt
 - decorations/
   - walls/
     - DarkOak/
       - Gate.nbt
       - ...

**Note 1:** schematics with multiple sub directories will have their style **directory/sub-directory/sub-sub-directory**. From the example above, Gate will have **walls/DarkOak**
**Note 2:** Multiple subdirectory work in the BuilderTool but this is not optimal as a long Style name will not fully show inside the tyle button. WindowBuildTool will need some love.

## Handle schematic location change:
### For building:
When a building style does not exist any more (because the directory have been renamed), the Building will use the md5 hash to find the new style:
Ex: 
- We have a Building *GuardTower* Level 1with style *wooden* (structure name: huts/wooden/TowerGuard1)
- The server administrator decided to rename the wooden directory to *OakWood* because he does have other wood type tower using different wood type.
- When loading the Building huts/wooden/TowerGuard1 the schematic is not found.
- We use the md5 hash to find that the schematic is now *huts/OakWood/TowerGuard1*
- Update the Building style to *OakWood*

### For a Work order:
For a work order, we do similar thing as we store the structure name and the md5. When the schematic is not found using the structure name, we use the md5 to find the new Schematic name and update the work order accordingly

**Note:** As the structure directory change prior the MD5 hash implementation some work order will fail after the upgrade because  the schematics file have changed. Those work order will be automatically cancelled (It will not be a problem for huts)

## ToDo
- Recover Building when schematic moved (done for work order)
- Clean up the code
- Add javadocs

xavierh committed:
Add a persistent UUID to the server, saved in minecolomies.dat
Send the UUID to the client when the player log in.
The UUID will be use to create a directory on the client to store data specific to a server

xavierh committed:
Send the Server UUID to the player
Register ServerUUIDMessage class

xavierh committed:
request as schematic from the client and send the schematic back by the server

xavierh committed:
Allow to override hits and decorations by nbt file in a directory
Fix problem with single player

xavierh committed:
Merge branch 'version/1.11' of https://github.com/Minecolonies/minecolonies into feature/Buildtool-server-side-nbt

marvin-bitterlich commented:
We could do an md5 checksum on the files to see if they differ

xavierh commented:
Yes that what I thought. I am going to do just that

On 12 Feb 2017 10:54, "Marvin Hofmann" <notifications@github.com> wrote:

> We could do an md5 checksum on the files to see if they differ
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Minecolonies/minecolonies/pull/579#issuecomment-279210391>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACplvEkJl8xYNFo-gW8Yc92GgdPlff1Cks5rbuTKgaJpZM4L-VAu>
> .
>

xavierh commented:
@Kostronor I should be able to work on that again this week.

I need to add the md5 checksum and will use **java.security.MessageDigest** unless you suggest something else.

Additionally I like to be able to create new style for huts not in the jar. For that I need  reorganize the directory structure for the schematics as huts and decorations are currently together.

I am aiming for:
- huts
  - wooden
    - builder1.nbt
    - builder2.nbt
    - ...
  - stone
    - builder1.nbt
    - builder2.nbt
    - ...
  - **My Own Style**
- decorations
  - decoration (I should probably change this name)
  - walls
  - waypoints


The idea behind it is that the directories in the **huts** directory are the styles, to create a new one, we just need to add a new directory with its nbt files in it.

I will need to pay attention to the current work orders so that any workorder using stone/builder1.nbt for instance will be converted to use huts/stone/builder1.nbt

/edit for formatting bullets list

marvin-bitterlich commented:
@xavierh 
That is a good idea. That was what we initially planned I guess :D

I suggest you just actively change it to actually have the huts part. That would make it way easier.

xavierh committed:
move hut schematics  in huts directory
move decoration schematic in decorations directory

xavierh committed:
Move the 2 last schematics
Generate the Server UUID after loading if not already set

xavierh committed:
Adjust huts and decorations location for the work orders

xavierh committed:
Transmit the schematic directly instead of using writeNBT so that the files or the asset from the jar are exactly the same (needed for MD5 check)

xavierh committed:
Check the MD5 hash when loadoing a schematic in the builder tools and request the schematic from the server when the MD5 hash does not match.

xavierh committed:
Save schematic in the correct folder

xavierh committed:
Postponed Structure.init, need a World Directory

xavierh committed:
No need to scan decoration in the gui

xavierh committed:
- Check the md5
- reorganise schematic handling and save directory

xavierh committed:
Merge branch 'version/1.11' of https://github.com/Minecolonies/minecolonies into feature/Buildtool-server-side-nbt

xavierh committed:
Fix merge issue with schematice location

xavierh committed:
Ignore decorations without a style
Organized Walls into sub directory (Oak, DarkOak, SandStone)

xavierh committed:
only huts in huts directory are build as a hut
load only from huts or decorations

xavierh commented:
TODO:
- recover work order when schematic have move: I will do that by adding the md5 hash to the work order. when a work order fail because the schematic does not exist anymore I will the use the md5 hash to see if the schematic location changed and update the work order accordingly.

- need to transition workorder pre pr to work order post pr. for hut it will be easy:
   if a hut structure is not found, we can look for huts/structurename if it does exit
- Transition will not be possible for decorations

Raycoms commented:
I would say we just cancel the decoration workOrders if we can't find them

Am Dienstag, 28. Februar 2017 schrieb xavierh :

> TODO:
>
>    -
>
>    recover work order when schematic have move: I will do that by adding
>    the md5 hash to the work order. when a work order fail because the
>    schematic does not exist anymore I will the use the md5 hash to see if the
>    schematic location changed and update the work order accordingly.
>    -
>
>    need to transition workorder pre pr to work order post pr. for hut it
>    will be easy:
>    if a hut structure is not found, we can look for huts/structurename if
>    it does exit
>    -
>
>    Transition will not be possible for decorations
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Minecolonies/minecolonies/pull/579#issuecomment-283046548>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGI9y1eX834WXktJZKB7hd7Mh0ThamyFks5rhCjogaJpZM4L-VAu>
> .
>

Raycoms commented:
In the past they crashed I think I've changed the code to canceling but not
sure if it works for all cases
Am Dienstag, 28. Februar 2017 schrieb Ray Neiheiser :

> I would say we just cancel the decoration workOrders if we can't find them
>
> Am Dienstag, 28. Februar 2017 schrieb xavierh :
>
>> TODO:
>>
>>    -
>>
>>    recover work order when schematic have move: I will do that by adding
>>    the md5 hash to the work order. when a work order fail because the
>>    schematic does not exist anymore I will the use the md5 hash to see if the
>>    schematic location changed and update the work order accordingly.
>>    -
>>
>>    need to transition workorder pre pr to work order post pr. for hut it
>>    will be easy:
>>    if a hut structure is not found, we can look for huts/structurename
>>    if it does exit
>>    -
>>
>>    Transition will not be possible for decorations
>>
>> —
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/Minecolonies/minecolonies/pull/579#issuecomment-283046548>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AGI9y1eX834WXktJZKB7hd7Mh0ThamyFks5rhCjogaJpZM4L-VAu>
>> .
>>
>

xavierh commented:
@Raycoms 
After the pr, all decoration structure name should start with decorations/
For instance **walls/stone**, should be **decorations/walls/stone** now.
If I see a Work Order Decoration where the structure is not found and the structurename does not start by **decorations/** then I will try to prefix it and see if I can load it then.

But I reorganize the Walls section as it was messy (https://github.com/Minecolonies/minecolonies/pull/579/commits/ab51c8c3d284b5bea77b0fa500a36b0f8d1c9fd4)

For example 
now we had:
decorations/walls/DarkOakWallCorner.nbt
which I change to:
decorations/walls/DarkOak/Corner.nbt 

So it is easier to navigate through walls schematic by style (**walls/DarkOak**).

Those one won't transition but I don't it is a big deal anyway.

The game does not crash when the structure of a work order does not exist. I tested it extensively.

Raycoms commented:
Okay that's important.
Yeah I had planned that but we didn't support subfolders yet

Am Dienstag, 28. Februar 2017 schrieb xavierh :

> @Raycoms <https://github.com/Raycoms>
> After the pr, all decoration structure name should start with decorations/
> For instance *walls/stone*, should be *decorations/walls/stone* now.
> If I see a Work Order Decoration where the structure is not found and the
> structurename does not start by *decorations/* then I will try to prefix
> it and see if I can load it then.
>
> But I reorganize the Walls section as it was messy (ab51c8c
> <https://github.com/Minecolonies/minecolonies/pull/579/commits/ab51c8c3d284b5bea77b0fa500a36b0f8d1c9fd4>
> )
>
> For example
> now we had:
> decorations/walls/DarkOakWallCorner.nbt
> which I change to:
> decorations/walls/DarkOak/Corner.nbt
>
> So it is easier to navigate through walls schematic by style (
> *walls/DarkOak*).
>
> Those one won't transition but I don't it is a big deal anyway.
>
> The game does not crash when the structure of a work order does not exist.
> I tested it extensively.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Minecolonies/minecolonies/pull/579#issuecomment-283051620>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGI9ywNJxZVH4_GL0rXDsBQdLzh0u_BGks5rhC2ZgaJpZM4L-VAu>
> .
>

xavierh committed:
Recover from moved schematic for building and work order using md5 hash

xavierh committed:
clean up

xavierh commented:
Ready for review

xavierh committed:
Fix crash

marchermans commented on src/main/java/com/minecolonies/coremod/colony/Structures.java:
```diff
@@ -25,23 +27,32 @@
+     * huts/stone/builder1.nbt -> hash
+     * decorations/decoration/Well.nbt -> hash
+     */
+    private static       Map<String, String>       md5Map                = new HashMap<>();
```
This can be improved by using a BiMap. Improves readability of reverse LoopUp and has the same performance

xavierh replied:
By BiMap do you mean com.google.common.collect.BiMap ?
Can I use this ?

In some case I can have 2 "different" schematics with the same hash.

I experienced this when I duplicate the TowerGuard schematic to do a corner version on.
Level 2,3,4,5 where different but Level 1 is identical as they is no door to place in the corner.

You are right, I will avoid mp5.isEmpty() and return null instead

marchermans replied:
Xavierh, if you have two schematics with the same hash, you are doing it wrong.

xavierh replied:
Why would it be wrong ?
The 2 schematics as identical

An other case where you could have twice the same schematic is when you want a bridge standard and a bridge which act as a waypoint.
both schematic would be identical, just the style (folder location) would be different.

I use MD5 to check the the client is using the same version of a schematic not to be sure they are unique.

Also I am planning for the player to be able to scan their own schematic, each player will have their own folder but schematic might be identical between to players

marchermans replied:
The idea behind you using these hashes is to find the differences between two files. So two different files (regardless of the fact that they belong to the same building, even if it is rotated) should never and i mean never have the same hash.

That would complete negate the principle of what you are working here.

On top of that, if you had two files with the same hash, your method to get a structure name from a hash would only return the first one. Which is maybe not the one you requested.

xavierh replied:
> The idea behind you using these hashes is to find the differences between two files. So two different files (regardless of the fact that they belong to the same building, even if it is rotated) should never and i mean never have the same hash.

No the idea is to make sure the client use the correct schematic.

> That would complete negate the principle of what you are working here.

No

>On top of that, if you had two files with the same hash, your method to get a structure name from a hash would only return the first one. Which is maybe not the one you requested.

One is better than none.

I use the hash to fix Work Order where the schematic still exist but have been renamed or moved. Yes this is not full proof but it is better that failing a WorkOrder and cancelling it. Or have a worthless building just because the schematic moved or its style name have been changed

marchermans replied:
I understand now what you are trying to achieve. Still i think we should prevent Hash Colisions at any cost. But that is my opinion.

Raycoms replied:
Why don't we enter the folder(style) and fileName into the hash? This would make it quite unique to avoid confusion.
If we want to prevent Hash colisions we shouldn't use md5 anyway. But Md5 is, for this use case, more than enough imo.

xavierh replied:
I (respectfully) disagree. using the filename and style in the MD5 would mean that we can not move the schematic around if needed.
Also it does not add any functionality to do it.

Raycoms replied:
the big question is. Do we need duplicated schematics?
What happens if we have duplicates on the server or on the client?
In the case of the guardTower corner, I think, the schematic which is the same for both styles should be more like a fallback than duplicated in both folders. (Check for equal md5 as required in own styles to check if we have it elsewhere then)
I see no reason to have duplicates around if we can define fallbacks or links.

xavierh replied:
The only reason why I had duplicated schematic is because I was testing the custom huts features and see if the clients pick up the new schematics.

I decided to go for the TowerGuard because I was in need to have a corner version.

I asked my builder to build the 5 different normal TowerGuards, one for each level.

Modified them to do the corner version. then scan them.

I did not rescan the level 1 because I did not change anything, and just copy the TowerGuard1 to the **Corner** style.

Also later one, player will be be able to scan. My plan is to store those in "schematics/players/${player UUID}/"
I can not control what player will scan if two players scan a 2x2x2 Cobblestone schematic, they will have the same MD5 hash. but each schematics will be in their own  player directory, so that each player can only uses their schematic not what another player have done.

>I see no reason to have duplicates around if we can define fallbacks or links.

We have nothing in place will allow you to have different path upgrade

marchermans commented on src/main/java/com/minecolonies/coremod/colony/Structures.java:
```diff
@@ -251,9 +292,81 @@ public static int getMaxLevelForHut(final String hut)
+     * @param decoration Decoration to get styles for.
+     * @return List of styles.
+     */
+    public static Map<String, String> getMD5s()
```
See comment to the md5 map variable.

marchermans commented on src/main/java/com/minecolonies/coremod/colony/Structures.java:
```diff
@@ -251,9 +292,81 @@ public static int getMaxLevelForHut(final String hut)
+     * @param md5 hash identifying the schematic
+     * @return the structure name as 'hut/wooden/Builder1' or an empty String if not found
+     */
+    public static String getStructureNameByMD5(final String md5)
```
By using a BiMap this method can be made a lot les complicated and a lot more performant.

marchermans commented on src/main/java/com/minecolonies/coremod/colony/Structures.java:
```diff
@@ -251,9 +292,81 @@ public static int getMaxLevelForHut(final String hut)
+     */
+    @SideOnly(Side.CLIENT)
+    public static void setMD5s(final Map<String, String> md5Map)
+    {
```
See comment for the md5Map field up above.

marchermans commented on src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java:
```diff
@@ -311,6 +317,26 @@ public void readFromNBT(@NotNull final NBTTagCompound compound)
+        final String md5 = compound.getString(TAG_SCHEMATIC_MD5);
+        final String structureName = Structures.SCHEMATICS_HUTS + '/' + style + '/' + this.getSchematicName() + buildingLevel;
+
+        if (Structures.getMD5(structureName).isEmpty())
```
It might be a good idea to add methods to the Structures class to check if it has an MD5 or a Structure Name.
This would make these arbitrary checks against isEmpty() not needed. Improve the readability and allow for more flexibility in the future, especially when creating an API for this kind of stuff, arbitrary checks should be avoided if they can simple be wrapped inside a proper method.

xavierh replied:
do you mean something like:
boolean hasMD5Hash(final String stuctureName) ?

fair point

marchermans commented on src/main/java/com/minecolonies/coremod/colony/workorders/WorkOrderBuild.java:
```diff
@@ -99,7 +100,23 @@ public void readFromNBT(@NotNull final NBTTagCompound compound)
         structureName = compound.getString(TAG_SCHEMATIC_NAME);
+        if (Structures.getMD5(structureName).isEmpty())
+        {
+            if (!Structures.getMD5(getStructurePrefix() + '/' + structureName).isEmpty())
```
Again see comment regarding line 325 of the AbstractBuilding class:

It might be a good idea to add methods to the Structures class to check if it has an MD5 or a Structure Name.
This would make these arbitrary checks against isEmpty() not needed. Improve the readability and allow for more flexibility in the future, especially when creating an API for this kind of stuff, arbitrary checks should be avoided if they can simple be wrapped inside a proper method.

marchermans commented on src/main/java/com/minecolonies/coremod/colony/workorders/WorkOrderBuild.java:
```diff
@@ -99,7 +100,23 @@ public void readFromNBT(@NotNull final NBTTagCompound compound)
         cleared = compound.getBoolean(TAG_IS_CLEARED);
+        md5 = compound.getString(TAG_SCHEMATIC_MD5);
         structureName = compound.getString(TAG_SCHEMATIC_NAME);
+        if (Structures.getMD5(structureName).isEmpty())
```
Again see comment regarding the AbstractBuilding class line 325:

It might be a good idea to add methods to the Structures class to check if it has an MD5 or a Structure Name.
This would make these arbitrary checks against isEmpty() not needed. Improve the readability and allow for more flexibility in the future, especially when creating an API for this kind of stuff, arbitrary checks should be avoided if they can simple be wrapped inside a proper method.

marchermans commented on src/main/java/com/minecolonies/coremod/network/messages/ColonyStylesMessage.java:
```diff
@@ -17,8 +17,10 @@
+    private Map<String, Integer> hutLevelsMap;
     private Map<String, List<String>> hutStyleMap;
     private Map<String, List<String>> decorationStyleMap;
+    private Map<String, String> md5Map;
```
If the change to a BiMap is made, it would be required here as well.

marchermans commented on src/main/java/com/minecolonies/coremod/network/messages/ColonyStylesMessage.java:
```diff
@@ -55,11 +76,39 @@ public void fromBytes(@NotNull final ByteBuf buf)
     }
 
+    @NotNull
+    private static Map<String, String> readMD5MapFromByteBuf(@NotNull final ByteBuf buf)
```
Again converting this to a final BiMap would improve performance and readability.

marchermans reviewed (commented):
I am leaving this review as general comment. However i suggest to @Kostronor and @Raycoms To look into these.

The BiMap change is easily made and improves the performance a lot. Especially since he is often querrying a Structure from a MD5 Hash.

On top of that with regards to the API, Voodoo-Checks like MD5.IsEmpty() should be avoided as no were in the documentation of MD5 is stated that it cannot produce an Empty String as Hash.
With regards to the API side of these things it would be a lot better if methods were added that for now just wrapped the isEmpty() checks for a given structure name, or md5 hashsum.

marchermans commented:
If it is the case, as stated above by @xavierh, that two files can have the same MD5 Hash i would not pull this, but find a different way of differentiating them.

xavierh committed:
add Structures.hasStructureName() method
Structures.getMD5 can now return null if structure not found

xavierh commented:
@OrionDevelopment 

As stated above, the standard use of the MD5 is to make sure client and server have the same schematic.

As a bonus, I use it to recover Work Order and Building when/if the files have moved.

I could have simply ignore this bit and let it fail (as the previous behaviour). Instead I choose to try to recover from moved/renamed schematic.

This is not a problem for decoration as it does not matter which one we use, the result will be the same.

For Building it is only a problem when two buildings are identical but their upgraded builing are different. In that case, yes, I might use the wrong style but it is still an improvement compare to the previous behaviour where you would have not be able to upgrade the building anymore.

marchermans commented:
I understand what you are trying to do, and i guess the case that the checksums are equal happens not so often (read so few times that is should not matter). 

Do i understand correctly that there could be two schematics with the same contents, used for different purposes?
So say one rotated and one not, if the contained building is a perfect mirror for all four horizontal axi?

If that is the case then yeah it would not matter and you cannot use a BiMap as that forces a one to one relationship.

However the problem still recides that a workorder that was rebuild from the MD5, is not capable of retrieving that second rotated (with the same MD5) structure, making it show the user the none rotated variant (granted the output of the workorder is equal, but still), so i donnot know if it is required but should be taken into account.

There are a lot of other ways of rebuilding the workorders and synchronizing the files, especially since we are storing them in NBT. However as for now, i think i can say that it should work.

xavierh commented:
> Do i understand correctly that there could be two schematics with the same contents, used for different purposes?

decoration with "waypoint" in the name are also use as a waypoint on top of being a decoration

> So say one rotated and one not, if the contained building is a perfect mirror for all four horizontal axi?

Rotation are not normally done via a different schematic, they are done when a work order is created, we set the position and rotation (and mirror soon) we want the builder to use.

>If that is the case then yeah it would not matter and you cannot use a BiMap as that forces a one to one relationship.

I don't understand what is the point of "forcing" the one to one relationship?

> However the problem still recides that a workorder that was rebuild from the MD5, is not capable of retrieving that second rotated (with the same MD5) structure, making it show the user the none rotated variant (granted the output of the workorder is equal, but still), so i donnot know if it is required but should be taken into account.

When the rotation is done inside the schematic, they will have different MD5 hash anyway.
When  the rotation is done within the work order, then any schematic we retrieve using the MD5 would do anyway (This is not the case for AbstactBuilding however)

marchermans commented:
Ah okey, i am understanding this now. Then you will not be able to fix your collision problem this way.

One last question before i shut up:
Is it possible for an other mod to override a given schematic?

xavierh commented:
> Ah okey, i am understanding this now. Then you will not be able to fix your collision problem this way.

I am not scared of collision. we use the structurename to identify the schematic not the md5. If we have a hash collision, that would only mean that the player could see a different build when using the BuildTool but the builder will use the server sided schematic.

> Is it possible for an other mod to override a given schematic?

Not cleanly. In theory,  a mod could dump new schematics in the WorldDir/minecolonies/schematics/ directory. but the list won't be refresh until the server restart.

marchermans commented:
@xavierh okey, then i am okey with this. Could you do me a favor. Directly under the class definition add a:
//TODO: [API]: Implement external API acces to Structures. Override and additional lookup.

That would be awesome.

xavierh commented:
I don't think this is the right place. The main branch to develop is normally **develop**.

This pr is for version/1.11 and will be back ported later to develop

marchermans commented:
I know, the API Stuff i am doing is for 1.11 and onwards. However tracking things like these when the person who created the PR does not take things like that into account is hard.

I made a note for it my self now.

But i might make a PR that adds twenty orso of those TODO's.

Raycoms commented on src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java:
```diff
@@ -397,15 +399,6 @@ private void hutDecClicked(@NotNull final Button button)
      */
     private List<String> getStyles()
     {
-        try
```
Does he still load the file onTheFly or do we want to relog to get the newest styles if we drop them off?

xavierh replied:
I removed this bit because it did work only in SP, this "scan" is only client side which mean, the player would have seen some schematic that the server is not aware of. The builder would not be able to build them anyway.

I will implement it properly with https://github.com/Minecolonies/minecolonies/issues/488 so it will work for SP with not restrictions and MP if allowed.

Raycoms replied:
Okay that would be nice. Its very nice if you can scan and then build on the go without having to restart.

xavierh replied:
Whatever method I will implement to add new schematic on the fly will trigger a refresh of this list.

xavierh replied:
@Raycoms You are right, 
I will load the client side directory but only when the worlds is not remote.
I guess I could do the same with the scan folder while I am at it ?

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/Structures.java:
```diff
@@ -123,28 +131,61 @@ public static void loadStyleMaps(final Path basePath) throws IOException
                     {
+                        if (isSchematicHut(hut))
+                        {
+                            Log.getLogger().warn(path + " look like a hut but will not be considered as a hut!");
```
That happens if someone drops of decorations in the hut folders?

xavierh replied:
This happen when a decoration schematic filename match a hut but is in decorations folder.
I added this warning to the log because the user might have put a hut schematic in the decorations folder without realising that it won't be working as a hut (but he will can still be able to build it).

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/Structures.java:
```diff
@@ -123,28 +131,61 @@ public static void loadStyleMaps(final Path basePath) throws IOException
                     }
+                    else
+                    {
+                        Log.getLogger().error(path + " schematic is not a huts or decorations, ignoring it");
```
that can happen? How?

xavierh replied:
I am scanning the whole **schematics** folder which is normally:
 - schematics
   - huts
   - decorations

There is nothing preventing the user to create more directory in schematics directory for instance **schematics/mystuff**. 
So in the log I wanted to make the user aware that it does not work.

However I need to change the Line 167:
`else if (SCHEMATICS_HUTS.equals(schematicCategory) || SCHEMATICS_DECORATIONS.equals(schematicCategory))`

for 
`else if (SCHEMATICS_DECORATIONS.equals(schematicCategory))`

because at the moment workOrderBuildDecoration assume that schematics are in "schematics/decorations"

This will also change in the futur when implementing https://github.com/Minecolonies/minecolonies/issues/488

Raycoms commented:
I'm currently thinking about pretty huge structures people might download
on the same website and will both upload.
The server could recognize he has that hash already and save space?
2017-03-02 11:15 GMT-03:00 xavierh <notifications@github.com>:

> *@xavierh* commented on this pull request.
> ------------------------------
>
> In src/main/java/com/minecolonies/coremod/colony/Structures.java
> <https://github.com/Minecolonies/minecolonies/pull/579#discussion_r103932076>
> :
>
> >      //Hut, Styles
>      private static       Map<String, List<String>> hutStyleMap           = new HashMap<>();
>      //Decoration, Style
>      private static       Map<String, List<String>> decorationStyleMap    = new HashMap<>();
>
> +    /* md5 hash for the schematics
> +     * format is:
> +     * huts/stone/builder1.nbt -> hash
> +     * decorations/decoration/Well.nbt -> hash
> +     */
> +    private static       Map<String, String>       md5Map                = new HashMap<>();
>
> The only reason why I had duplicated schematic is because I was testing
> the custom huts features and see if the clients pick up the new schematics.
>
> I decided to go for the TowerGuard because I was in need to have a corner
> version.
>
> I asked my builder to build the 5 different normal TowerGuards, one for
> each level.
>
> Modified them to do the corner version. then scan them.
>
> I did not rescan the level 1 because I did not change anything, and just
> copy the TowerGuard1 to the *Corner* style.
>
> Also later one, player will be be able to scan. My plan is to store those
> in "schematics/players/${player UUID}/"
> I can not control what player will scan if two players scan a 2x2x2
> Cobblestone schematic, they will have the same MD5 hash. but each
> schematics will be in their own player directory, so that each player can
> only uses their schematic not what another player have done.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Minecolonies/minecolonies/pull/579#discussion_r103932076>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGI9y5NHfeIWBMh_qHiUmJ46FNUk9Gm0ks5rhs77gaJpZM4L-VAu>
> .
>

xavierh commented:
@Raycoms I understand your point but this PR does not allow to upload stuff, that is why I did not bother about those considerations.

Changes will be made when required I guess.

Raycoms commented:
:shipit: for me then.

[![Approved with PullApprove](https://img.shields.io/badge/maintainers-approved-brightgreen.svg)](https://pullapprove.com/Minecolonies/minecolonies/pull-request/579/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

Raycoms commented:
Can you create a backport to 1.10 as well?

xavierh committed:
decorations in the huts folder should be ignored as the work order will not work with that anyway

xavierh commented:
@Raycoms As discussed I changed Line 167

For 1.10, what is the best way to backport, could you briefly describe the process with git ?

Raycoms commented:
I usually make a squash merge in another 1.11 branch. Then cherrypick this
in a copy of develop of 1.10 and then I fix the compile errors and test it.

2017-03-02 11:44 GMT-03:00 xavierh <notifications@github.com>:

> @Raycoms <https://github.com/Raycoms> As discussed I changed Line 167
>
> For 1.10, what is the best way to backport, could you briefly describe the
> process with git ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Minecolonies/minecolonies/pull/579#issuecomment-283672021>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGI9ywYyWnELtUHb9sRZEtDN6NFXQh7Pks5rhtXjgaJpZM4L-VAu>
> .
>

xavierh commented:
@Raycoms 
cool new stuff to learn. Would it not make send to squash now so it would be only one commit in version/1.11 ?

Raycoms commented:
We will have to wait for @Marvin or @colton to approve your pr to squash
merge already. And that might need a while =D

2017-03-02 12:08 GMT-03:00 xavierh <notifications@github.com>:

> @Raycoms <https://github.com/Raycoms>
> cool new stuff to learn. Would it not make send to squash now so it would
> be only one commit in version/1.11 ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Minecolonies/minecolonies/pull/579#issuecomment-283678962>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGI9y77ezTWA3I-GYHu9vhbXFZc40xVAks5rhttygaJpZM4L-VAu>
> .
>

cltnschlosser commented:
Didn't we decide we didn't want to force users to download schematics. They can be made available through resource packs, and if you don't have the pack, you just won't see a preview in the build tool.

marvin-bitterlich commented:
@cltnschlosser not force maybe, but at least offer to do it

marvin-bitterlich commented:
@xavierh I suggest you ping is somewhere when you change the pr status (and request our review ;) )

xavierh commented:
@Kostronor Was I that slow ? :D

marvin-bitterlich commented:
@xavierh It was sitting here for quite some time. If there wasn't ray asking me to recheck it I'd just had ignored it 😛

marvin-bitterlich commented:
Oh btq @xavierh I guess we have to move the pr anyway because sonar etc. are not working on forks. Would that be okay for you?

xavierh commented:
Yes
I would like someone to have a second close look to a specific part, please

marvin-bitterlich commented:
It's 108 files, I won't review that before sonar has a go at it xD ping me on specific things you want answered :)

Raycoms commented:
We can, shouldn't be a problem.

2017-03-02 19:13 GMT-03:00 xavierh <notifications@github.com>:

> *@xavierh* commented on this pull request.
> ------------------------------
>
> In src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java
> <https://github.com/Minecolonies/minecolonies/pull/579#discussion_r104043411>
> :
>
> > @@ -397,15 +399,6 @@ private void hutDecClicked(@NotNull final Button button)
>       */
>      private List<String> getStyles()
>      {
> -        try
>
> @Raycoms <https://github.com/Raycoms> You are right,
> I will load the client side directory but only when the worlds is not
> remote.
> I guess I could do the same with the scan folder while I am at it ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Minecolonies/minecolonies/pull/579#discussion_r104043411>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGI9y23QlegoJ_lFq3is6J0GAtN-NSW7ks5rhz77gaJpZM4L-VAu>
> .
>

xavierh commented:
I need to work on this a bit more. as @Raycoms mention, I lost a functionality.
SP player can not drop schematic in the decoration folder and use it **without restarting** the game.

xavierh commented:
Could somebody comment on those lines:
https://github.com/Minecolonies/minecolonies/blob/feature/Buildtool-server-side-nbt/src/main/java/com/minecolonies/structures/helpers/Structure.java#L163

xavierh committed:
Fix crash when md5 does not exist

marchermans commented:
@xavierh Is that not always true cause you are using the directory as a prefix?

xavierh commented:
I think not.
The client could ask for a schematic named "decorations/../../anyfile"
The question is, does this test is enough ?

I should also do it on the client side so when the client download a file he does not write outside of its directory

xavierh committed:
Fix security issue: prevent writig file outside of the schematics directory

xavierh commented:
Todo:

- add scan folder back (the location will change) for sp user
- save cached  schematic using the md5 hash instead to avoid security issue. (Also will be reused for player custom schematic on mp)

Unfortunately I will need more time too impment it correctly :(

xavierh committed:
Add custom decoration (from local player)
Cache file using the md5 hash as a file name inside cache directory
basically we have 4 paths now:
huts: only for customs huts
decorations: decoration available on the server
cache: use by the client to cache schematic from the server or by the server when player use their own schematics (decoration only)
custom: custom decoration schematic on the client only

When the player want to use custom, he will need to send the schematic unless the cache/<md5 hash> already exists

Schematic can be send to the server only if enabled in the configuration (disable by default), no schematic management is done yet, no limitation on the number of schamatic send, unused one are still there.

xavierh committed:
Reimplement WindowBuildTool
Still need cleaning and disable custom schematic on servers

xavierh committed:
cleaning up

xavierh committed:
cleaning up + javadoc

xavierh committed:
more cleaning

xavierh committed:
more cleanup

xavierh committed:
Add John Reider parks
http://www.minecolonies.com/publicforum/m/18051739/viewthread/29772689-for-your-consideration-town-park/

xavierh committed:
Use a human readable name for stared/finished message for custom decoration

xavierh committed:
Fix Null Pointer exception with decoration and construction tape

xavierh committed:
cleanup

xavierh committed:
Build schematicsMap on the client from the md5Map, saving bandwith and processing on the server

xavierh committed:
Cleanup schematic loading

xavierh committed:
cleanup

xavierh committed:
add md5 in cache when sending to server

xavierh committed:
rename hasStructureName to hasMD5

xavierh committed:
Merge branch 'version/1.11' of https://github.com/Minecolonies/minecolonies into feature/Buildtool-server-side-nbt

xavierh committed:
Create a new constructor for Structurename(section, style, schematic)
Fix getStructureNameByMD5

xavierh committed:
create missing directory for additionnal schematics

xavierh committed:
Send schematic list to players whenever the list change

xavierh committed:
Check that the server accept shematic before sending

xavierh commented:
Will create a clean pr again when ready
